dot/parachain/dispute/backend.go

```diff
@@ -0,0 +1,207 @@
+package dispute
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/parachain/dispute/types"
+	parachainTypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/dgraph-io/badger/v4"
+	"github.com/tidwall/btree"
+)
+
+// Backend is the backend for the dispute coordinator module.
+type Backend interface {
+	// GetEarliestSession returns the earliest session index, if any.
+	GetEarliestSession() (*parachainTypes.SessionIndex, error)
+	// GetRecentDisputes returns the recent disputes, if any.
+	GetRecentDisputes() (*btree.BTree, error)
+	// GetCandidateVotes returns the votes for the given candidate for the specific session-candidate pair, if any.
+	GetCandidateVotes(session parachainTypes.SessionIndex, candidateHash common.Hash) (*types.CandidateVotes, error)
+
+	// SetEarliestSession sets the earliest session index.
+	SetEarliestSession(session *parachainTypes.SessionIndex) error
+	// SetRecentDisputes sets the recent disputes.
+	SetRecentDisputes(recentDisputes *btree.BTree) error
+	// SetCandidateVotes sets the votes for the given candidate for the specific session-candidate pair.
+	SetCandidateVotes(session parachainTypes.SessionIndex, candidateHash common.Hash, votes *types.CandidateVotes) error
+}
+
+// OverlayBackend is the overlay backend for the dispute coordinator module.
+type OverlayBackend interface {
+	Backend
+
+	// WriteToDB writes the given dispute to the database.
+	WriteToDB() error
+	// GetActiveDisputes returns the active disputes.
+	GetActiveDisputes(now int64) (*btree.BTree, error)
+}
+
+// DBBackend is the backend for the dispute coordinator module that uses a database.
+type DBBackend interface {
+	Backend
+
+	// Write writes the given data to the database.
+	Write(earliestSession *parachainTypes.SessionIndex,
+		recentDisputes *btree.BTree,
+		candidateVotes map[types.Comparator]*types.CandidateVotes) error
+}
+
+type syncedEarliestSession struct {
+	*sync.RWMutex
```

Comment:
I would suggest using a property name to hold those types instead of embedding them in the struct

```go
type syncedEarliestSession struct {
   mtx sync.RWMutex
   sessionIndex *parachainTypes.SessionIndex
}
```

Reply:
Could you elaborate on why you'd prefer it that way?

Reply:
After some research, I couldn't find any real constraint for using this implicit composition style. Looks like you want to have those structures concurrently safe since most of them use `sync.RWMutex` so I would suggest including the prefix/suffix `sync` or `shared` (similar to what they use in rust for structures that need to be protected against concurrent read/write). Also, I would suggest to use the `sync.RWMutext` value rather than the pointer to it

Reply:
I think this change would fix any deep source issues with embedding as well

Reply:
Sorry, been meaning to ping Eclesio on this; been a crazy couple of weeks. 

@EclesioMeloJunior I don't understand, doesn't it already include the prefix?

Reply:
@kanishkatn nevermind I didn't pay attention it was already as the prefix 😅

Reply:
Yeah aside from deciding whether to embed or not, I am not sure of why its better to have a pointer to the mutex. 

My personal opinion is if I pointer is not needed then do not use one, though others may disagree here. Is there any benefit here to using a pointer?

Reply:
https://github.com/ChainSafe/gossamer/pull/3344#discussion_r1328508507

I'll resolve this!